### PR TITLE
Fix udp socket

### PR DESCRIPTION
**Pull request by hannesm**

### CODE REVIEW

**Comment by hannesm** [comment]
@mirage/core this came out of using stack-socket. unfortunately SO_REUSEPORT is only available from OCaml 4.12 onwards -- anyone has problems with dropping earlier OCaml releases in tcpip?

**Comment by djs55** [comment]
I'm personally happy with OCaml 4.12+

File: src/stack-unix/udpv4_socket.ml [hunk 1]
```diff
@@ -30,6 +30,7 @@ let get_udpv4_listening_fd {listen_fds;interface} port =
     Lwt.return @@ Hashtbl.find listen_fds (interface,port)
   with Not_found ->
     let fd = Lwt_unix.(socket PF_INET SOCK_DGRAM 0) in
+    Lwt_unix.(setsockopt fd SO_REUSEPORT true);
```
Comment (djs55) [inline src/stack-unix/udpv4_socket.ml:33]: Is this to allow multiple Unix processes to bind to the same port, to share responsibility for handling the requests?

Comment (hannesm) [reply src/stack-unix/udpv4_socket.ml:33]: This is to allow the same unix process to bind to the same port several times, i.e. in DNS:
- we listen on UDP Ipaddr_any, 53 (via a listen callback) -- calls `bind(socket, *, 53)`
- once a UDP frame is received, we take some time to find an answer, and eventually send _from_ (*, 53) to the client -- this means we have to bind as well to (*, 53).

Without the REUSEPORT we can't bind to the same ip, port pair multiple times.

**Comment by dinosaure** [comment]
> @mirage/core this came out of using stack-socket. unfortunately SO_REUSEPORT is only available from OCaml 4.12 onwards -- anyone has problems with dropping earlier OCaml releases in tcpip?

Hmmhmm, 4.12 is a bit higher for me. We should keep, at least, the support of the 3 last version of OCaml I think.

**Comment by hannesm** [comment]
> 4.12 is a bit higher for me

ok, so what are the options?
- leave tcpip broken as is (the socket stack can't receive on the same port as it wants to send on (via UDP))
- provide a C stub that allows to set SO_REUSEPORT
- rework this PR to be able to avoid SO_REUSEPORT (the commit https://github.com/mirage/mirage-tcpip/commit/76b857821c1af5d4bfd603b0df790e04489d4159#diff-2df233b86746a505609a07125cd5a1ca625c8ddc4faba8ab44ddf9c2a0d9b578 introduced the `create_socket` asymmetry in Udpv4v6_socket) -- the `create_socket` exists since when sending UDP data we only want to use a single socket (and the get_udpv4v6_listening_fds may return two sockets (one for IPv4, one for IPv6 if both a static v4 and a v6 address is configured). Likely this can be refactored.

Eventually, the SO_REUSEPORT is not required if -- similar to how Udpv4_socket (and Udpv6_socket) re-use listening sockets for writing. Unfortunately the lifetime of a socket is then infinite (with file descriptors leakage) -- I guess the stack-socket needs some enhancements anyways (and there should be the counterpart to llsten_tcp/listen_udp that stops the listener).

> 3 last version of OCaml

From my memory, we agreed to the last 2 OCaml versions -- and 4.13 is upcoming soon. I also think that the tcpip stack is barely used by others than MirageOS unikernels. IMHO it is fine to bump to 4.12 (and 4.13) for this library.

So:
- this PR here fixes some issues if you want to use UDP and the socket stack, esp. with dual stacks
- there's more work needed to avoid leaking file descriptors
- it is not urgent since it seems like nobody really uses the socket stack and UDP (we use it for developing and testing a dns resolver, but we can work with a local pin for now)

**Comment by hannesm** [comment]
I opened #450 describing the leaks and potential fix(es). related to #446. I also moved the less controversial part to #451.

**Comment by dinosaure** [comment]
> From my memory, we agreed to the last 2 OCaml versions -- and 4.13 is upcoming soon. I also think that the tcpip stack is barely used by others than MirageOS unikernels. IMHO it is fine to bump to 4.12 (and 4.13) for this library.

So if we agree with that, it should fine so :+1:. I largely prefer a non-buggy stack.

**Review by djs55 — APPROVED** [review]

**Comment by avsm** [comment]
A minimum of 4.12 is just fine by me for this package

**Comment by hannesm** [comment]
we revised the implementation to address the file descriptor leaks (should solve #446 / #450).
also, we removed the REUSEPORT socket option (which was really only needed in udpv4v6_socket), and instead reuse the same socket for listen and write (by adding some more logic into get_udpv4v6_listening_fd).

TL;DR: the socket stack `disconnect` function now cleanup file descriptors. the tests suite should pass nicely, and we don't require OCaml 4.12. we tested this with the dns-resolver unikernel which reads from and writes to a lot of udp sockets.

any reviews welcome, I'm keen to merge and release a new version of tcpip without the leaks (plus the other changes in previous PRs).
